lib/pengine/pe_actions.c

```diff
@@ -74,71 +74,112 @@ find_existing_action(const char *key, const pcmk_resource_t *rsc,
+ *
+ * \return XML configuration of desired action if any, otherwise NULL
+ */
 static xmlNode *
-find_rsc_op_entry_helper(const pcmk_resource_t *rsc, const char *key,
-                         gboolean include_disabled)
+find_exact_action_config(const pcmk_resource_t *rsc, const char *key,
+                         bool include_disabled)
 {
-    guint interval_ms = 0;
-    gboolean do_retry = TRUE;
-    char *local_key = NULL;
-    const char *name = NULL;
-    const char *interval_spec = NULL;
-    char *match_key = NULL;
-    xmlNode *op = NULL;
-    xmlNode *operation = NULL;
+    xmlNode *action_config = NULL;
 
-  retry:
-    for (operation = pcmk__xe_first_child(rsc->ops_xml); operation != NULL;
-         operation = pcmk__xe_next(operation)) {
+    for (xmlNode *operation = first_named_child(rsc->ops_xml, XML_ATTR_OP);
+         operation != NULL; operation = crm_next_same_xml(operation)) {
 
-        if (pcmk__str_eq((const char *)operation->name, "op", pcmk__str_none)) {
-            bool enabled = false;
+        bool enabled = false;
+        guint interval_ms = 0;
+        char *match_key = NULL;
+        const char *name = crm_element_value(operation, "name");
+        const char *interval_spec = crm_element_value(operation,
+                                                      XML_LRM_ATTR_INTERVAL);
 
-            name = crm_element_value(operation, "name");
-            interval_spec = crm_element_value(operation, XML_LRM_ATTR_INTERVAL);
-            if (!include_disabled && pcmk__xe_get_bool_attr(operation, "enabled", &enabled) == pcmk_rc_ok &&
-                !enabled) {
-                continue;
-            }
+        // @TODO This does not consider rules, defaults, etc.
+        if (!include_disabled
+            && (pcmk__xe_get_bool_attr(operation, "enabled",
+                                       &enabled) == pcmk_rc_ok) && !enabled) {
+            continue;
+        }
 
-            interval_ms = crm_parse_interval_spec(interval_spec);
-            match_key = pcmk__op_key(rsc->id, name, interval_ms);
-            if (pcmk__str_eq(key, match_key, pcmk__str_casei)) {
-                op = operation;
-            }
-            free(match_key);
+        interval_ms = crm_parse_interval_spec(interval_spec);
 
-            if (rsc->clone_name) {
-                match_key = pcmk__op_key(rsc->clone_name, name, interval_ms);
-                if (pcmk__str_eq(key, match_key, pcmk__str_casei)) {
-                    op = operation;
-                }
-                free(match_key);
-            }
+        // Try first with resource ID
+        match_key = pcmk__op_key(rsc->id, name, interval_ms);
+        if (pcmk__str_eq(key, match_key, pcmk__str_none)) {
+            action_config = operation;
+        }
+        free(match_key);
 
-            if (op != NULL) {
-                free(local_key);
-                return op;
+        // Then again with clone_name in case ID has instance number
```

Comment:
Shouldn't we turn this around (haven't checked yet if you did later) for efficiency and skip over the 2nd search if we found one?

Reply:
Some callers of `pcmk__find_action_config()` pass a key with the resource (instance) ID and some with the history (clone) ID. There doesn't appear to be a clear winner, so I'll leave it.